src/librustc/mir/visit.rs

```diff
@@ -880,125 +945,156 @@ pub enum TyContext {
 }
 
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
-pub enum PlaceContext<'tcx> {
-    // Appears as LHS of an assignment
-    Store,
+pub enum NonMutatingUseContext {
+    /// Being inspected in some way, like loading a len.
+    Inspect,
+    /// Consumed as part of an operand.
+    Copy,
+    /// Consumed as part of an operand.
+    Move,
+}
 
-    // Can often be treated as a Store, but needs to be separate because
-    // ASM is allowed to read outputs as well, so a Store-AsmOutput sequence
-    // cannot be simplified the way a Store-Store can be.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum MutatingUseContext {
+    /// Appears as LHS of an assignment.
+    Store,
+    /// Can often be treated as a `Store`, but needs to be separate because
+    /// ASM is allowed to read outputs as well, so a `Store`-`AsmOutput` sequence
+    /// cannot be simplified the way a `Store`-`Store` can be.
     AsmOutput,
-
-    // Dest of a call
+    /// Destination of a call.
     Call,
-
-    // Being dropped
+    /// Being dropped.
     Drop,
+}
 
-    // Being inspected in some way, like loading a len
-    Inspect,
-
-    // Being borrowed
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum MaybeMutatingUseContext<'tcx> {
+    /// Being borrowed.
     Borrow { region: Region<'tcx>, kind: BorrowKind },
 
-    // Used as base for another place, e.g. `x` in `x.y`.
-    //
-    // The `Mutability` argument specifies whether the projection is being performed in order to
-    // (potentially) mutate the place. For example, the projection `x.y` is marked as a mutation
-    // in these cases:
-    //
-    //     x.y = ...;
-    //     f(&mut x.y);
-    //
-    // But not in these cases:
-    //
-    //     z = x.y;
-    //     f(&x.y);
+    /// Used as base for another place, e.g. `x` in `x.y`.
+    ///
+    /// The `Mutability` argument specifies whether the projection is being performed in order to
+    /// (potentially) mutate the place. For example, the projection `x.y` is marked as a mutation
+    /// in these cases:
+    ///
+    ///     x.y = ...;
+    ///     f(&mut x.y);
+    ///
+    /// But not in these cases:
+    ///
+    ///     z = x.y;
+    ///     f(&x.y);
     Projection(Mutability),
+}
 
-    // Consumed as part of an operand
-    Copy,
-    Move,
-
-    // Starting and ending a storage live range
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum NonUseContext {
+    /// Starting a storage live range.
     StorageLive,
+    /// Ending a storage live range.
     StorageDead,
-
-    // Validation command
+    /// User type annotation assertions for NLL.
+    AscribeUserTy,
+    /// Validation command.
     Validate,
 }
 
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum PlaceContext<'tcx> {
+    NonMutatingUse(NonMutatingUseContext),
+    MutatingUse(MutatingUseContext),
+    MaybeMutatingUse(MaybeMutatingUseContext<'tcx>),
```

Comment:
This is not nice... now we have a mixture of type-based classification and dynamic classification, not getting the benefits of either of them.

Given this, I don't think it makes any sense to separate the enum into 4.

Reply:
I've modified this to remove the fourth variant. It's still not ideal, because there are now three borrow variants, but I think it is cleaner than with four variants.